jxl/src/features/patches.rs

```diff
@@ -0,0 +1,290 @@
+            return Err(Error::PatchesTooMany(num_ref_patch, max_ref_patches));
+        }
+        let mut total_patches = 0;
+        let mut next_size = 1;
+        let mut positions: Vec<PatchPosition> = Vec::new();
+        let mut blendings = Vec::new();
+        let mut ref_positions: Vec<PatchReferencePosition> = Vec::with_capacity(num_ref_patch);
+        for _ in 0..num_ref_patch {
+            let reference = patches_reader.read(br, REFERENCE_FRAME_CONTEXT)? as usize;
+            if reference >= DecoderState::MAX_STORED_FRAMES {
+                return Err(Error::PatchesRefTooLarge(
+                    reference,
+                    DecoderState::MAX_STORED_FRAMES,
+                ));
+            }
+
+            let x0 = patches_reader.read(br, PATCH_REFERENCE_POSITION_CONTEXT)? as usize;
+            let y0 = patches_reader.read(br, PATCH_REFERENCE_POSITION_CONTEXT)? as usize;
+            let ref_pos_xsize = patches_reader.read(br, PATCH_SIZE_CONTEXT)? as usize + 1;
+            let ref_pos_ysize = patches_reader.read(br, PATCH_SIZE_CONTEXT)? as usize + 1;
+            let reference_frame = decoder_state.reference_frame(reference);
+            match reference_frame {
+                None => return Err(Error::PatchesInvalidReference(reference)),
+                Some(reference) => {
+                    if !reference.saved_before_color_transform {
+                        return Err(Error::PatchesPostColorTransform());
+                    }
+                    if x0 + ref_pos_xsize > reference.frame[0].size.0 {
+                        return Err(Error::PatchesInvalidPosition(
+                            "x".to_string(),
+                            x0,
+                            ref_pos_xsize,
+                            reference.frame[0].size.0,
+                        ));
+                    }
+                    if y0 + ref_pos_ysize > reference.frame[0].size.1 {
+                        return Err(Error::PatchesInvalidPosition(
+                            "y".to_string(),
+                            y0,
+                            ref_pos_ysize,
+                            reference.frame[0].size.1,
+                        ));
+                    }
+                }
+            }
+
+            let id_count = patches_reader.read(br, PATCH_COUNT_CONTEXT)? as usize + 1;
+            if id_count > max_patches + 1 {
+                return Err(Error::PatchesTooMany(id_count, max_patches));
+            }
+            total_patches += id_count;
+
+            if total_patches > max_patches {
+                return Err(Error::PatchesTooMany(total_patches, max_patches));
+            }
+
+            if next_size < total_patches {
+                next_size *= 2;
+                next_size = std::cmp::min(next_size, max_patches);
+            }
+            positions.reserve(next_size);
+            blendings.reserve(next_size * PATCH_BLEND_MODE_NUM_BLEND_MODES as usize);
```

Comment:
note that reserve() takes as argument the *additional* space to ensure, not the total size...

Reply:
Hah, then sometimes the calls in libjxl are reserving less than the capacity of the vector size, which then has no effect. Not quite sure if this is intended. I think in any case asymptotically this makes sense. We could do exacly the same here by doing 
```
if next_size > positions.len() {
    positions.reserve(next_size - positions.len());
}

```
Or, and I did this now, because I think it makes sense, to add `1` to `next_size`, everytime we push to positions. Then we don't need the check because as an invariant `next_size` will be larger than `positions.len`verytime we push to positions.
Please check that this works, or I can just go back to exactly what libjxl does, which does not quite make sense to me.

Reply:
Could also do `positions.try_reserve(next_size.saturating_sub(positions.len())`, I guess..

Reply:
ok, doing that now...